test/vagrant-service-manager/proxy_test.rb

```diff
@@ -0,0 +1,109 @@
+      end
+
+      describe 'Unauthenticated proxy environment' do
+        before do
+          @string_stream = StringIO.new
+          @options = { type: :docker, '--cli-version' => '1.10.0' }
+          @handler = ADBDockerBinaryHandler.new(@machine, @machine.env, @options)
+        end
+
+        after do
+          Thread.kill @download_thread
+          Thread.kill @proxy_thread
+          @proxy_server.shutdown
+        end
+
+        it 'should send client binary download request' do
+          @handler.build_download_url
+          @handler.validate_url
+          @handler.build_archive_path
+          @handler.ensure_binary_and_temp_directories
+
+          @proxy_thread = Thread.new do
+            @configs[:port] = 7000
+            @proxy_server = fake_proxy_server(@handler.url, @string_stream, @configs)
+            @proxy_server.start
+          end
+          @machine.config.servicemanager.proxy = 'http://localhost:7000'
+
+          begin
+            @download_thread = Thread.new do
+              @handler.download_archive
+            end
+          rescue StandardError => e
+            raise e.message
+          end
+
+          @string_stream.truncate(0)
+          sleep 3 # Require for @download_thread to start request
+
+          expected_result = "CONNECT get.docker.com:443 200 URL=#{@handler.url}"
+          @string_stream.string.must_include expected_result
+        end
+      end
+
+      describe 'Authenticated proxy environment' do
+        before do
+          @string_stream = StringIO.new
+          @options = { type: :docker, '--cli-version' => '1.10.0' }
+          @handler = ADBDockerBinaryHandler.new(@machine, @machine.env, @options)
+        end
+
+        after do
+          Thread.kill @download_thread
+          Thread.kill @proxy_thread
+          @proxy_server.shutdown
+        end
+
+        it 'should send client binary download request' do
+          @handler.build_download_url
+          @handler.validate_url
+          @handler.build_archive_path
+          @handler.ensure_binary_and_temp_directories
+
+          @proxy_thread = Thread.new do
+            @configs[:port] = 7001
+            @configs[:authenticated] = true
+            @proxy_server = fake_proxy_server(@handler.url, @string_stream, @configs)
+            @proxy_server.start
+          end
+          @machine.config.servicemanager.proxy = 'http://localhost:7001'
+          @machine.config.servicemanager.proxy_user = 'user'
+          @machine.config.servicemanager.proxy_password = 'password'
+
+          begin
+            @download_thread = Thread.new do
+              @handler.download_archive
+            end
+          rescue StandardError => e
+            raise e.message
+          end
+
+          sleep 3 # Require for @download_thread to start request
+
+          expected_result = "CONNECT get.docker.com:443 200 URL=#{@handler.url}"
+          @string_stream.string.must_include expected_result
```

Comment:
This is all a bit convoluted. Why do we not try to encapsulate this all a bit better by pushing more functionality into the proxy. So when we create the proxy, we pass it a list of expected URLs. Then the proxy creates its own StringIO instance and we push as much of the config side of things and thread management into this as well. this makes the actual test leaner as well.

Last but not least we have a method on the proxy which is of the form _proxy.assertRequestReceived_ which internally will go through the logs and verify the requests are in there.

Moving forward one can then add things like _proxy.assertRequestReceivedInOrder_ etc. The fake proxy stuff should do as much as possible.